### PR TITLE
[Collective][PR 4/6] NCCL Communicator caching and preliminary stream management

### PR DESCRIPTION
**Pull request by zhisbug**
This is the *4th PR* for the project [Collective-in-Ray](https://github.com/ray-project/ray/pull/12174). 

See a list below:

1.  PR #12637 ( **merged**) ~~Basic infrastructure; an in-actor collective interface `ray.util.collective.init_collective_group(*args, **kwargs)`; support for two collectives `allreduce` and `barrier`; some testing infrastructure, etc.~~
2. (PR #12874 **merged**) ~~Driver-program interface, which includes: (1) the second interface: `actor.options(collective_options, ...).remote()` and the third interface `declare_collective_group(actors, collective_options, ...)`.~~
3. (PR #12864 **merged**) ~~Support for other collectives: `allgather`, `broadcast`, `reduce`, `reducescatter`; refactor the tests into distributed tests and single-node cluster tests~~. 
3.5 (PR #12935 **merged**) ~~send/recv P2P communication apis.~~ 
4. (**this one**) Communicator caching,  and support for num_gpus > 2 per actor/task.
5. CUDA stream management.
6. docs, examples, etc.

## Related issue number
See this [rfc-collective-in-ray](https://github.com/ray-project/RFC/blob/main/rfc-20201119-collective-in-ray/20201119-collective-in-ray.md)

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by zhisbug** [comment]
@richardliaw This is a big one but good to go! Now it supports any multi-GPU collective calls on actors with >1 gpus!

File: python/ray/util/collective/tests/sinlge_node_tests/test_basic_apis.py [hunk 1]
```diff
@@ -65,7 +65,7 @@ def test_availability(ray_start_single_node_2_gpus):
         actors[0].report_nccl_availability.remote())
     assert actor0_nccl_availability
     actor0_mpi_availability = ray.get(
-        actors[0].report_mpi_availability.remote())
+        actors[0].report_gloo_availability.remote())
```
Comment (richardliaw) [inline python/ray/util/collective/tests/sinlge_node_tests/test_basic_apis.py:68]: is this right? should you fix the var name?

File: python/ray/util/collective/collective_group/nccl_util.py [hunk 2]
```diff
@@ -194,6 +200,25 @@ def get_tensor_strides(tensor):
...
 
 
+def get_tensor_device(tensor):
+    """Return the GPU index of a tensor."""
+    if isinstance(tensor, cupy.ndarray):
+        try:
+            device = tensor.device.id
+        except AttributeError as e:
+            raise RuntimeError("Got an exception {}. The tensor is not on "
+                               "a valid GPU.".format(e))
```
Comment (richardliaw) [inline python/ray/util/collective/collective_group/nccl_util.py:210]: `raise RuntimeError() from ..`

File: python/ray/util/collective/collective_group/nccl_util.py [hunk 3]
```diff
@@ -228,3 +253,23 @@ def copy_tensor(dst_tensor, src_tensor):
...
+        list: the list of GPU devices.
+
+    """
+    if not isinstance(tensors, list):
+        raise RuntimeError(
+            "Expect a list of tensors each locates on a GPU device. "
+            "Got: '{}'.".format(type(tensors)))
+    devices = [0] * len(tensors)
+    for i, tensor in enumerate(tensors):
+        devices[i] = get_tensor_device(tensor)
```
Comment (richardliaw) [inline python/ray/util/collective/collective_group/nccl_util.py:274]: devices = [get_tensor_device(t) for t in tensors]

File: python/ray/util/collective/collective_group/nccl_util.py [hunk 4]
```diff
@@ -194,6 +200,25 @@ def get_tensor_strides(tensor):
...
+        try:
+            device = tensor.device.id
+        except AttributeError as e:
+            raise RuntimeError("Got an exception {}. The tensor is not on "
+                               "a valid GPU.".format(e))
+    elif torch_available():
+        if isinstance(tensor, torch.Tensor):
+            device = tensor.device.index
+            if not isinstance(device, int):
+                raise RuntimeError("The tensor is not on a valid GPU.")
```
Comment (richardliaw) [inline python/ray/util/collective/collective_group/nccl_util.py:215]: I think you need to merge the tensor and torch_available check (or else you have a unbounded local variable error)

File: python/ray/util/collective/tests/conftest.py [hunk 5]
```diff
@@ -1,30 +1,41 @@
...
+import logging
 
+import pytest
 import ray
+from ray.util.collective.collective_group.nccl_collective_group \
+    import _get_comm_key_from_devices, _get_comm_key_send_recv
 from ray.util.collective.const import get_nccl_store_name
 
+logger = logging.getLogger(__name__)
+logger.setLevel("INFO")
```
Comment (richardliaw) [inline python/ray/util/collective/tests/conftest.py:11]: hmm just to be safe, can you remove the setLevel?

File: python/ray/util/collective/collective_group/nccl_collective_group.py [hunk 6]
```diff
@@ -31,13 +27,18 @@ class Rendezvous:
     process.
 
     Args:
-        group_name (str): the unique user-specified group name.
+        store_key (str): the unique store key, usually as a concatanation
+                         of group_name and communicator key.
+                         See `get_nccl_communicator` for more details.
```
Comment (richardliaw) [inline python/ray/util/collective/collective_group/nccl_collective_group.py:32]: Indent once -- 

```suggestion
        store_key (str): the unique store key, usually as a concatenation
            of group_name and communicator key.
            See `get_nccl_communicator` for more details.
```

File: python/ray/util/collective/collective.py [hunk 7]
```diff
@@ -7,28 +7,24 @@
...
-#     import MPIGroup
-# except ImportError:
-#     _MPI_AVAILABLE = False
 try:
     from ray.util.collective.collective_group import NCCLGroup
 except ImportError:
     _NCCL_AVAILABLE = False
 
 logger = logging.getLogger(__name__)
+logger.setLevel("INFO")
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:19]: remove?

File: python/ray/util/collective/collective.py [hunk 8]
```diff
@@ -206,9 +214,8 @@ def get_world_size(group_name: str = "default") -> int:
         group_name: the name of the group to query
 
     Returns:
-        The world size of the collective group，
-        -1 if the group does not exist or the process does
-        not belong to the group.
+        The world size of the collective group, -1 if the group does
+        not exist or the process does not belong to the group.
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:218]:
```suggestion
        The world size of the collective group, -1 if the group does
            not exist or the process does not belong to the group.
```

File: python/ray/util/collective/collective.py [hunk 9]
```diff
@@ -232,7 +239,29 @@ def allreduce(tensor, group_name: str = "default", op=types.ReduceOp.SUM):
...
     opts = types.AllReduceOptions
     opts.reduceOp = op
-    g.allreduce(tensor, opts)
+    g.allreduce([tensor], opts)
+
+
+def allreduce_multigpu(tensor_list: list,
+                       group_name: str = "default",
+                       op=types.ReduceOp.SUM):
+    """Collective allrecue a list of tensors across the group.
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:248]:
```suggestion
    """Collective allreduce a list of tensors across the group.
```

File: python/ray/util/collective/collective.py [hunk 10]
```diff
@@ -232,7 +239,29 @@ def allreduce(tensor, group_name: str = "default", op=types.ReduceOp.SUM):
...
+
+    Args:
+        tensor_list (List[tensor]): list of tensors to be allreduced,
+                                    each on a GPU.
+        group_name (str): the collective group name to perform allreduce.
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:253]:
```suggestion
        tensor_list (List[tensor]): list of tensors to be allreduced,
             each on a GPU.
        group_name (str): the collective group name to perform allreduce.
```

File: python/ray/util/collective/collective.py [hunk 11]
```diff
@@ -271,7 +300,42 @@ def reduce(tensor,
...
+                    dst_rank: int = 0,
+                    dst_tensor: int = 0,
+                    group_name: str = "default",
+                    op=types.ReduceOp.SUM):
+    """Reduce the tensor across the group to the destination rank
+    and destination tensor.
+
+    Args:
+        tensor_list: the list of tensors to be reduced on this process;
+                     each tensor located on a GPU.
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:317]:
```suggestion
        tensor_list: the list of tensors to be reduced on this process;
             each tensor located on a GPU.
```

File: python/ray/util/collective/collective.py [hunk 12]
```diff
@@ -346,11 +464,38 @@ def reducescatter(tensor,
...
+                           input_tensor_lists,
+                           group_name: str = "default",
+                           op=types.ReduceOp.SUM):
+    """Reducescatter a list of tensors across all GPUs.
+
+    Args:
+        output_tensor_list: the resulted list of tensors, with
+                            shape: num_gpus * shape(tensor).
+        input_tensor_lists: the original tensors, with shape:
+                            num_gpus * world_size * shape(tensor).
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:480]:
```suggestion
        output_tensor_list: the resulted list of tensors, with
            shape: num_gpus * shape(tensor).
        input_tensor_lists: the original tensors, with shape:
            num_gpus * world_size * shape(tensor).
```

File: python/ray/util/collective/collective.py [hunk 13]
```diff
@@ -462,7 +675,7 @@ def _check_rank_valid(g, rank: int):
     """Check the rank: 0 <= rank < world_size."""
     if rank < 0:
         raise ValueError("rank '{}' is negative.".format(rank))
-    if rank > g.world_size:
+    if rank >= g.world_size:
         raise ValueError("rank '{}' is greater than world size "
                          "'{}'".format(rank, g.world_size))
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:680]:
```suggestion
        raise ValueError("rank '{}' must be less than world size "
                         "'{}'".format(rank, g.world_size))
```

File: python/ray/util/collective/collective.py [hunk 14]
```diff
@@ -476,3 +689,24 @@ def _check_tensor_list_input(tensor_list):
...
         _check_single_tensor_input(t)
+
+
+def _check_tensor_lists_input(tensor_lists):
+    """Check if the input is a list of lists of supported tensor types."""
+    if not isinstance(tensor_lists, list):
+        raise RuntimeError("The input must be a list of lists of tensors. "
+                           "Got '{}'.".format(type(tensor_lists)))
+    if not tensor_lists:
+        raise RuntimeError("Got an empty list of lists of tensors.")
```
Comment (richardliaw) [inline python/ray/util/collective/collective.py:700]:
```suggestion
    if not tensor_lists:
        raise RuntimeError(f"Did not receive tensors. Got: {tensor_lists}")
```

**Review by richardliaw — CHANGES_REQUESTED** [review]
Still need to review `nccl_collective_group`. Will review tomorrow!

File: python/ray/util/collective/tests/sinlge_node_tests/test_basic_apis.py (hunk 1 above)
Comment (zhisbug) [reply python/ray/util/collective/tests/sinlge_node_tests/test_basic_apis.py:68]: yes you're 100% correct! 👍

**Comment by zhisbug** [comment]
> Still need to review `nccl_collective_group`. Will review tomorrow!

Addressed!

**Review by richardliaw — APPROVED** [review]

**Pull request merged**